sql: add hash join algorithm for USING/NATURAL predicates

irfansharif opened the pull request:
Implementation of the hash join algorithm, limited to only `NATURAL` and
`USING` predicates. `ON` predicates with explicit equality constraints 
(i.e. `a.x = b.x`), at least for now, fall back to the nested loop join implementation.

There's an intermediary result buffer attached to the join node here,
this cleans up the code compared to an earlier implementation with
maintained state for consecutive calls to `Next()` and `Values()` and can be
used similarly for other plan nodes where the computation of a batch of
results is simpler to reason about with any 'state maintenance' pushed down to
the buffer itself.

Due to #10534 the current buffer implementation is non-functional,
extending `RowContainer` to support deletes at the front now to address this.
(edit: #10623, done)

Addresses #10397.

cc @RaduBerinde @knz

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10618)
<!-- Reviewable:end -->

knz commented:
Seeing how the join tests needed to be changed to use ORDER BY made me smile; it's exactly what SQL is supposed to be!
Thanks for working on this.

1) Regarding the first commit, please split further between one commit with the rename and one with the change. Otherwise I can't see the change.

2) In general avoid moving methods around in the same commit where you change them. This obscures the changes and makes review excessively difficult.

3) Change ExplainPlan to report the algorithm; then add new EXPLAIN tests that show how different algorithms are used for different join types.

4) You're being a bit too liberal with RowContainer. I think this can be improved.

---

Reviewed 3 of 4 files at r1, 2 of 3 files at r2, 11 of 11 files at r3.
Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[pkg/sql/join.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI0GlTOtF0LaYSrmQN:-KWI0GlUjsbP5HnKlkto:b-x6ct7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f053403009208291b4605f1d27420dbd4b3a340b/pkg/sql/join.go#L43)):*
> ```Go
> )
> 
> // commonColumns returns the names of columns common on the
> ```

Just curious, why did you pull this method (commonColumns) up so far from the constructor that uses it?

Oh then, I see in the next commit it disappears from here again.  
Could you ensure that this change disappears from the intermediate commit diff? So that anyone inspecting the changes later does not get distracted.

---

*[pkg/sql/join.go, line 34 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI8KOFoAvyHZAFx2yw:-KWI8KOFoAvyHZAFx2yx:b2nwvac) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L34)):*
> ```Go
> const (
> 	joinTypeInner joinType = iota
> 	joinTypeLeftOuter
> ```

If you think it is important to rename these constants, please split in a different commit. Right now it obscures the diff.

---

*[pkg/sql/join.go, line 47 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI16bng99izsVIvK47:-KWI16bng99izsVIvK48:b-1tjd2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L47)):*
> ```Go
> // there was a matching group (with the same group key) in the opposite stream.
> type bucket struct {
> 	rows *RowContainer
> ```

Use a single RowContainer for the entire joinNode, then take slices of its row array in each bucket.
Use a single MemoryAccount for the entire joinNode to account for the slice container.

Separating RowContainers is only meaningful if they would have (say) different lists of columns. 
You only want to Close() a single container; Close()ing is (relatively) expensive; you don't want to call it thousands of times.

---

*[pkg/sql/join.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI1fxB_GvDHDiLRt-M:-KWI1fxChwOAkgHWbprv:bg2es14) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L116)):*
> ```Go
> }
> 
> // makeJoin constructs a planDataSource for a JOIN node.
> ```

Don't move this function away from where it was originally defined in the same commit where you modify it. Right now in the diff I see a diff for the entire function (move), and the particular changes you've made are obscured by this. This makes it very hard to review.

---

*[pkg/sql/join.go, line 313 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI2Bk7FS4v8YTIq-tz:-KWI2Bk8CIqX5FU8DE7a:btu1gb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L313)):*
> ```Go
> 				n.rightRows = nil
> 			}
> 		} else if n.joinAlgorithm == hashJoin {
> ```

Split the code for the different loops in different functions; also use switch not if else.

---

*[pkg/sql/join.go, line 390 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI2Rz59ipztxJk5Gjp:-KWI2Rz6snUkXvGfAc8M:b-kfrgjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L390)):*
> ```Go
> 		return n.debugNext()
> 	}
> 	if n.joinAlgorithm == nestedLoopJoin {
> ```

switch

---

*[pkg/sql/join.go, line 397 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI2TGPHzcU3mQ87OUv:-KWI2TGPHzcU3mQ87OUw:biuftxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L397)):*
> ```Go
> 	}
> 
> 	return false, errors.New("unsupported JOIN algorithm")
> ```

extract the error into a global err variable.

---

*[pkg/sql/join.go, line 543 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI326_JwZKA87obpiS:-KWI326_JwZKA87obpiT:b1an1sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L543)):*
> ```Go
> 		// order to use it for RIGHT OUTER joins but if we encounter another
> 		// NULL row when going through the left stream (probing phase), matching
> 		// this with the first NULL row would be incorrect.
> ```

Extend this explanatory comment to highlight what happens when joining on 2 or more columns, and only one of them contains NULL.

---

*[pkg/sql/join.go, line 576 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI43gpG6AIMjG71va2:-KWI43gqk0bSDiW8pWka:b-1diun4) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L576)):*
> ```Go
> 		if !ok {
> 			if n.joinType == joinTypeInner {
> 				scratch = encoded[:0]
> ```

Add a comment here "// Failed to match -- no matching row, nothing to do."

---

*[pkg/sql/join.go, line 579 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI4ADThUYX9B3luCpu:-KWI4ADULkp0PQcCBjh8:b7k9sni) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L579)):*
> ```Go
> 				continue
> 			}
> 			// Given we didn't find a matching right row we append an empty
> ```

Insert a comment line above here "// Outer join: unmatched rows are padded with NULLs."

---

*[pkg/sql/join.go, line 641 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5GzUgrr0AVK-oLRQ:-KWI5GzV_yF72IzWdac_:b-kfrgjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L641)):*
> ```Go
> // Values implements the planNode interface.
> func (n *joinNode) Values() parser.DTuple {
> 	if n.joinAlgorithm == nestedLoopJoin {
> ```

switch

---

*[pkg/sql/join.go, line 663 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5Iy9e5XN4LSu2sMN:-KWI5Iy9e5XN4LSu2sMO:b-kfrgjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L663)):*
> ```Go
> // Close implements the planNode interface.
> func (n *joinNode) Close() {
> 	if n.joinAlgorithm == nestedLoopJoin {
> ```

switch

---

*[pkg/sql/join.go, line 673 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5KNPnMvt-Bybshst:-KWI5KNPnMvt-Bybshsu:b-18394t) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L673)):*
> ```Go
> 		n.buffer.Close()
> 		n.buffer = nil
> 		for _, b := range n.buckets {
> ```

See my comment above; this is insane.

As a general rule of thumb, the complexity of Close() for an entire logical plan should be linear in the number of nodes of the logical plan, and fully independent of the data in the database. That's the entire design contract of the monitor/account machinery. If we didn't want that, we may as well skip MemoryAccount entirely and use a loop in RowContainer to substract consumed memory row by row.

---

*[pkg/sql/join_predicate.go, line 37 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5v6qz9r2cTlx00H0:-KWI5v6rzs6YJEVhDR_U:b-awdije) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join_predicate.go#L37)):*
> ```Go
> 	// input data sources.
> 	prepareRow(result, leftRow, rightRow parser.DTuple)
> 	encode([]byte, parser.DTuple, string) ([]byte, bool, error)
> ```

1) Add an explanatory comment for this new method.

2) don't use a string in the 3rd argument position, instead use an int with aptly named constants in the global scope.

---

*[pkg/sql/join_predicate.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI6HBu0DV630nyEomZ:-KWI6HBu0DV630nyEom_:bf46b6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join_predicate.go#L206)):*
> ```Go
> func (p *usingPredicate) encode(b []byte, row parser.DTuple, side string) ([]byte, bool, error) {
> 	var cols []int
> 	if side == "right" {
> ```

Use int for "side"; also use switch not if here.

---

*[pkg/sql/join_predicate.go, line 285 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI6bwflFmI2YnzxMUP:-KWI6bwflFmI2YnzxMUQ:bjufi1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join_predicate.go#L285)):*
> ```Go
> 		if parser.ReNormalizeName(col.Name) == colName {
> 			idx = j
> 			break
> ```

Woah good catch, but not the right fix. There should be a check here before the assignment to `idx`: `if idx != invalidColIdx { return 0, parser.TypeNull, fmt.Errorf("USING: column name %s is ambiguous in %s table", colName, context) }`.

---

*[pkg/sql/row_buffer.go, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI7K1W0Da99KxBOwPF:-KWI7K1Xyxv5iHYiKRF7:bsw78so) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/row_buffer.go#L33)):*
> ```Go
> type RowBuffer struct {
> 	rows   *RowContainer
> 	output parser.DTuple
> ```

I'd run PopFirst() at the beginning of Next() and avoid `output` entirely (use only `At` in `Values()`).

---

*[pkg/sql/distsql/hashjoiner.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI7ZnqJZGkdYHHMS25:-KWI7ZnqJZGkdYHHMS26:b-osg0v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/distsql/hashjoiner.go#L147)):*
> ```Go
> 				return err
> 			}
> 			if row != nil && !h.output.PushRow(row) {
> ```

This seems unrelated. Please extract in a separate commit (and probably PR).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWI-oz09QKYXQg9lklb:b-5vxioz)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/sql/join.go, line 97 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI8gphMo-Nx0v-kpIm:-KWI8gpiZRk3m8BsdnXB:bxx1oer) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L97)):*
> ```Go
> 
> 	buckets map[string]bucket
> 	acc     mon.BoundAccount
> ```

Don't use BoundAccount here; use WrappableMemoryAccount. See `group.go` for examples.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618)*
<!-- Sent from Reviewable.io -->

irfansharif committed:
sql: refactor join predicates

refactor out join predicate definitions, implementations into separate
file.

irfansharif committed:
sql: implement hash join, NATURAL/USING predicates

Implementation of the hash join algorithm, limited to only NATURAL and
USING predicates. ON predicates, atleast for now, fall back to the
nested join implementation.

There's an intermediary result buffer attached to the join node here,
this cleans up the code compared to an earlier implementation with
maintained state for consecutive calls to Next() and Values() and can be
used similarly for other plan nodes where the computation of a batch of
results is simpler to reason about with state maintenance pushed down to
the buffer itself.

irfansharif committed:
sql: better constant names, correct camel casing

irfansharif commented:
> 1) Regarding the first commit, please split further between one commit with the rename and one with the change. Otherwise I can't see the change.

Done.

> 2) In general avoid moving methods around in the same commit where you change them. This obscures the changes and makes review excessively difficult.

that's a good tip, thank you! restructured and moved commits around to reflect this now.

---

Review status: 1 of 9 files reviewed at latest revision, 19 unresolved discussions.

---

*[pkg/sql/join.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI0GlTOtF0LaYSrmQN:-KWIvx18IRG5dmhShxO_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f053403009208291b4605f1d27420dbd4b3a340b/pkg/sql/join.go#L43)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Just curious, why did you pull this method (commonColumns) up so far from the constructor that uses it?
> 
> Oh then, I see in the next commit it disappears from here again.  
> Could you ensure that this change disappears from the intermediate commit diff? So that anyone inspecting the changes later does not get distracted.
</details>
Done.

---

*[pkg/sql/join.go, line 34 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI8KOFoAvyHZAFx2yw:-KWIvveYySEWpzt0NvlR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L34)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> If you think it is important to rename these constants, please split in a different commit. Right now it obscures the diff.
</details>
Done.

---

*[pkg/sql/join.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI1fxB_GvDHDiLRt-M:-KWIvzvrTeN5XlSmGHkk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L116)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Don't move this function away from where it was originally defined in the same commit where you modify it. Right now in the diff I see a diff for the entire function (move), and the particular changes you've made are obscured by this. This makes it very hard to review.
</details>
Done.

---

*[pkg/sql/distsql/hashjoiner.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI7ZnqJZGkdYHHMS25:-KWIw7H_Qol0hoOtzOpm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/distsql/hashjoiner.go#L147)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This seems unrelated. Please extract in a separate commit (and probably PR).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWIvdAy7JbrA8_EpyzS:bihiplm)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Review status: 1 of 9 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/sql/join.go, line 313 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI2Bk7FS4v8YTIq-tz:-KWYbPfFmQ1oj2BeASbA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L313)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Split the code for the different loops in different functions; also use switch not if else.
</details>
Done.

---

*[pkg/sql/join.go, line 390 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI2Rz59ipztxJk5Gjp:-KWYbV083hTHmr-iJSpf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L390)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> switch
</details>
Done.

---

*[pkg/sql/join.go, line 397 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI2TGPHzcU3mQ87OUv:-KWYbXoVdxiBjMhUSUAI:bqs40ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L397)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> extract the error into a global err variable.
</details>
changed this to a `panic` instead.

---

*[pkg/sql/join.go, line 543 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI326_JwZKA87obpiS:-KWZ-gZpgKvVXGZQ6bTa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L543)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Extend this explanatory comment to highlight what happens when joining on 2 or more columns, and only one of them contains NULL.
</details>
Done.

---

*[pkg/sql/join.go, line 576 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI43gpG6AIMjG71va2:-KWYhd4yAyBaqGfBH4Hb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L576)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Add a comment here "// Failed to match -- no matching row, nothing to do."
</details>
Done.

---

*[pkg/sql/join.go, line 579 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI4ADThUYX9B3luCpu:-KWYhu5ZeHSy5K4SobCB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L579)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Insert a comment line above here "// Outer join: unmatched rows are padded with NULLs."
</details>
Done.

---

*[pkg/sql/join.go, line 641 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5GzUgrr0AVK-oLRQ:-KWYbjIEB5J1DrgzyLLE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L641)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> switch
</details>
Done.

---

*[pkg/sql/join.go, line 663 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5Iy9e5XN4LSu2sMN:-KWYbmh2P9Zblcu5mUs1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L663)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> switch
</details>
Done.

---

*[pkg/sql/row_buffer.go, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI7K1W0Da99KxBOwPF:-KWZ07hPBA0937dNsSPw:b-zaszol) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/row_buffer.go#L33)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'd run PopFirst() at the beginning of Next() and avoid `output` entirely (use only `At` in `Values()`).
</details>
`PopFirst()` discards the first row from the internal `RowContainer`, we have to retain a reference to the original discarded row. It makes more intuitive sense for me to _discard_ a row when calling `Next()` as opposed to `Values()`, fits in better with the rest of the codebase. 

---

*[pkg/sql/join_predicate.go, line 37 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5v6qz9r2cTlx00H0:-KWYfsO-W1AlGN6TOR4z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join_predicate.go#L37)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> 1) Add an explanatory comment for this new method.
> 
> 2) don't use a string in the 3rd argument position, instead use an int with aptly named constants in the global scope.
</details>
Done.

---

*[pkg/sql/join_predicate.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI6HBu0DV630nyEomZ:-KWYfvUGgAesMBkJo9xm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join_predicate.go#L206)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Use int for "side"; also use switch not if here.
</details>
Done.

---

*[pkg/sql/join_predicate.go, line 285 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI6bwflFmI2YnzxMUP:-KWYgMu0eBHTnQw8GZHG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join_predicate.go#L285)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Woah good catch, but not the right fix. There should be a check here before the assignment to `idx`: `if idx != invalidColIdx { return 0, parser.TypeNull, fmt.Errorf("USING: column name %s is ambiguous in %s table", colName, context) }`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for addressing most of the concerns. The topic of the row containers is still open though.

---

Reviewed 2 of 11 files at r3, 2 of 9 files at r4, 2 of 2 files at r5, 3 of 4 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/row_buffer.go, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI7K1W0Da99KxBOwPF:-KWZwlj0UrhM8wvHt7ZV:b2nhll9) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/row_buffer.go#L33)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary>
> `PopFirst()` discards the first row from the internal `RowContainer`, we have to retain a reference to the original discarded row. It makes more intuitive sense for me to _discard_ a row when calling `Next()` as opposed to `Values()`, fits in better with the rest of the codebase. 
</details>
Ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWZxAWeP4wPs9W-K1C8:bhh57g2)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
For the first commit, did you use `git mv`? It would be important for history purposes. The diffs (both on github and reviewable) look like they didn't realize the relationship. You might need to do two commits, one that moves `join to index_join` and the second one that moves `table_join` to `join`

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_join.go, line 241 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KW_-wjnA1A9KrirfsTa:-KW_-wjnA1A9KrirfsTb:bimudic) ([raw file](https://github.com/cockroachdb/cockroach/blob/928ccd5333a4ad94da2e83cbe4ae6089c1b98e28/pkg/sql/index_join.go#L241)):*
> ```Go
> 		return true, nil
> 	}
> 	return false, nil
> ```

This refactoring is not correct; the loop is needed - what if there are no rows produced from this batch? (`n.table` can have a filter).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWZzziplGIEAOEC-hYA:bklspaj)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_join.go, line 241 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KW_-wjnA1A9KrirfsTa:-KW_2Psb047cHvOqflCo:b-n45hv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/928ccd5333a4ad94da2e83cbe4ae6089c1b98e28/pkg/sql/index_join.go#L241)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> This refactoring is not correct; the loop is needed - what if there are no rows produced from this batch? (`n.table` can have a filter).
</details>
You might be able to trigger with existing tests by temporarily setting indexJoinBatchSize to something small (1?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618)*
<!-- Sent from Reviewable.io -->

irfansharif committed:
sql: Collate all uses of the memory account

Shared RowContainer between all buckets, separate memory account used
for RowBuffer structure.
Additionally embedded RowContainer into RowBuffer to avoid duplicate
AddRow() and Close() functionality.

irfansharif committed:
sql: sql/join.go moved to sql/index_join.go

renamed file to match the rest of the package layout.

irfansharif committed:
sql: sql/table_join.go moved to sql/join.go

renamed file to match the rest of the package layout.

irfansharif commented:
@RaduBerinde: I did use `git mv` but I think `join.go` being common to both operations threw things off. Separated.

---

Review status: 0 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_join.go, line 241 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KW_-wjnA1A9KrirfsTa:-KW_BtKI3jZcF2iloMV7:b9j3yj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/928ccd5333a4ad94da2e83cbe4ae6089c1b98e28/pkg/sql/index_join.go#L241)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> You might be able to trigger with existing tests by temporarily setting indexJoinBatchSize to something small (1?)
</details>
ah. Removed.

---

*[pkg/sql/join.go, line 47 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI16bng99izsVIvK47:-KWZSZIwXFE7-cVE99f2:b-b4p6zm) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L47)):*
> Use a single RowContainer for the entire joinNode, then take slices of its row array in each bucket.

This is not implemented currently as per discussions offline.

> Use a single MemoryAccount for the entire joinNode to account for the slice container. You only want to Close() a single container; Close()ing is (relatively) expensive; you don't want to call it thousands of times.
 
Done

---

*[pkg/sql/join.go, line 97 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI8gphMo-Nx0v-kpIm:-KW_4GkhvpunfRSaFYcM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L97)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Don't use BoundAccount here; use WrappableMemoryAccount. See `group.go` for examples.
</details>
Done.

---

*[pkg/sql/join.go, line 673 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI5KNPnMvt-Bybshst:-KW_6XuG9RnhPpc6goSC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L673)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> See my comment above; this is insane.
> 
> As a general rule of thumb, the complexity of Close() for an entire logical plan should be linear in the number of nodes of the logical plan, and fully independent of the data in the database. That's the entire design contract of the monitor/account machinery. If we didn't want that, we may as well skip MemoryAccount entirely and use a loop in RowContainer to substract consumed memory row by row.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KW_64Q8qWI40bll83zI:b2akfzz)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
@knz: PTAL, latest revision demonstrates my ideas about the use of a shared `RowContainer` across all hash join buckets. Quite simply, use the RowContainer to **actually** store the rows, buckets simply store references to the stored copy in the `RowContainer`.

---

Review status: 0 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/join.go, line 47 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWI16bng99izsVIvK47:-KWdBuiYqdsAadgRDXAh:bc9em1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c05c5d17972b08c476db60dc2ccf4bb0fb5a0c/pkg/sql/join.go#L47)):*
> Use a single RowContainer for the entire joinNode, then take slices of its row array in each bucket.

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWdBNC8liL0Eslc9wIk:btm37pr)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
> 3) Change ExplainPlan to report the algorithm; then add new EXPLAIN tests that show how different algorithms are used for different join types.

Done.

---

Review status: 0 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWhoCAK6P7YwUlqG_zQ:bvxo11w)*
<!-- Sent from Reviewable.io -->

knz commented:
I have started the review from scratch. Thanks to you splitting the changes in different commits, the experience was much more pleasant this time!

The overall changes LGTM. The overall architecture of join looks fine this way.

However you did something with the memory accounts which I rather strongly disagree with. A lot of the current complexity in the code was added to avoid interfaces in the monitoring API and their call overhead, and you've just introduced one.

So I'd like us to look at this together and explore alternatives. If you wish I can also attempt to explore this myself and submit a PR towards your branch. Whatever you're most comfortable with.

---

Reviewed 2 of 11 files at r3, 8 of 9 files at r9, 2 of 2 files at r10, 2 of 2 files at r11, 3 of 3 files at r12, 5 of 5 files at r13.
Review status: 6 of 19 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/join.go, line 162 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWi2m1Ku481gqRVOzxi:-KWi2m1Lsn_G7RTW9sI1:bwilh25) ([raw file](https://github.com/cockroachdb/cockroach/blob/89f4384957f13a3bafa67c7090fb6d229eee787f/pkg/sql/join.go#L162)):*
> ```Go
> 		case *parser.UsingJoinCond:
> 			pred, info, err = p.makeUsingPredicate(leftInfo, rightInfo, t.Cols)
> 		default:
> ```

So why did you remove the `default` case here?
("here" = `makeJoin` / `switch t := cond.(type)`)

---

*[pkg/sql/row_buffer.go, line 31 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWi44-Xtj5PQiaCVOua:-KWi44-Yc1FJGiIp7Lfh:b-qd4czu) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8bc0643005b7ab76130fd55d5e06a113cb22f9/pkg/sql/row_buffer.go#L31)):*
> ```Go
> // under the constraints imposed by Next() and Values() under the planNode
> // interface.
> type RowBuffer struct {
> ```

I think if you make RowBuffer inherit from *RowContainer you don't have to implement the two methods Close and AddRow.

---

*[pkg/sql/row_container.go, line 31 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWi5OilJzZpW0BrSbXE:-KWi5OilJzZpW0BrSbXF:b-uiwqcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c890561a1f617671a8fbec63d9eafcbf5fb5739/pkg/sql/row_container.go#L31)):*
> ```Go
> // TODO(irfansharif): Push this down to sql/mon. Collate differences between
> // WrappedMemoryAccount and mon.BoundMonitor, etc.
> type memoryMonitor interface {
> ```

No, please don't go there. I do not wish us to pay the price of an interface for this use case. :)
There must be a better way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWi5pXojpDDBGyzRo3D:bj2zvqv)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
I investigated possible alternatives here some more, I want to bring back to question what exactly do we have to gain using a `WrappedMemoryAccount` as opposed to a `mon.BoundAccount` that is still `txnState` bound. We seem to be using the latter in our nested loop join implementation as well (see `newContainerValuesNode` in `values.go`) and this retains the original implementation on `RowContainer` working off a `mon.BoundAccount`. 

Additionally (possibly related) do you foresee the unification of the two (`WrappedMemoryAccount` and `mon.BoundAccount`) as used in `sql`? Currently I see both usages throughout and it's not immediately apparent to me what benefits we get from having two separate ways to maintain an account?

---

Review status: 6 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/join.go, line 162 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWi2m1Ku481gqRVOzxi:-KWiKPi9BTIHXEr_nMEp:be1fj0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/89f4384957f13a3bafa67c7090fb6d229eee787f/pkg/sql/join.go#L162)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> So why did you remove the `default` case here?
> ("here" = `makeJoin` / `switch t := cond.(type)`)
</details>
```
default:
	err = util.UnimplementedWithIssueErrorf(2970, "unsupported JOIN predicate: %T", cond)
```

it used to be this, my understanding was we've covered all JOIN predicates and the issue was therefore subsequently closed. am I missing something?

---

*[pkg/sql/row_buffer.go, line 31 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWi44-Xtj5PQiaCVOua:-KWiJj4q0HnodGtqzq61:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8bc0643005b7ab76130fd55d5e06a113cb22f9/pkg/sql/row_buffer.go#L31)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I think if you make RowBuffer inherit from *RowContainer you don't have to implement the two methods Close and AddRow.
</details>
Done.

---

*[pkg/sql/row_container.go, line 31 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWi5OilJzZpW0BrSbXE:-KWiJq0PUf6SkmcgTCdE:b-11qipo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c890561a1f617671a8fbec63d9eafcbf5fb5739/pkg/sql/row_container.go#L31)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> No, please don't go there. I do not wish us to pay the price of an interface for this use case. :)
> There must be a better way.
</details>
hmm, I hadn't thought of the cost of indirection here. Need to revise this. Look at the top level comment above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWijDdDIo04aaRiA-1u:b-vgkhvg)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Oh I see, cleaner transitioning between session specific and transaction specific contexts. I'm still not sure `WrappedMemoryAccount` is a better fit in this case however (as compared to `mon.BoundAccount` i.e.).

---

Review status: 6 of 20 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWiteG5TtHCPcR9fxH_:b-vy9al)*
<!-- Sent from Reviewable.io -->

irfansharif committed:
sql: tracking mem usage of bucket row references

irfansharif commented:
@knz: PTAL, reverted back to the original use of `mon.BoundMonitor`. Additionally please look at the last commit in isolation, it covers all the changes necessary to track memory usage for storing all the `parser.DTuple` references, this forces me to pass around `s *Session` adding on to some of the simpler APIs I had originally wanted for `buckets` but no strong opinions on this.

---

Review status: 6 of 19 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWjR7RHQRG4oYr5ghYq:b-xn07wo)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, modulo a suggestion for readability improvement on the last commit.

---

Reviewed 1 of 2 files at r10, 14 of 14 files at r15, 2 of 2 files at r16.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/join.go, line 162 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWi2m1Ku481gqRVOzxi:-KWmt4jLUVWK7_8dhNd4:b-h0rd5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/89f4384957f13a3bafa67c7090fb6d229eee787f/pkg/sql/join.go#L162)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary>
> ```
> default:
> 	err = util.UnimplementedWithIssueErrorf(2970, "unsupported JOIN predicate: %T", cond)
> ```
> 
> it used to be this, my understanding was we've covered all JOIN predicates and the issue was therefore subsequently closed. am I missing something?
</details>
It was a protection against future changes: if someone ever adds a new condition type but forgets to do something about here, it will be caught properly instead of doing something non-sensical in the constructor below.
But it's not very important either. Your call.

---

*[pkg/sql/join.go, line 76 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWn1K9CJZSqpUNP8k9R:-KWn1K9CJZSqpUNP8k9S:b945t0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d74aa090b918f268b3b2705f7f8bb9fd86b31ad9/pkg/sql/join.go#L76)):*
> ```Go
> }
> 
> func (b *buckets) AddRow(s *Session, encoding []byte, row parser.DTuple) error {
> ```

- Add the WrappableMemoryAccount in joinNode.
- pass WrappedMemoryAccount as argument to `buckets.AddRow`, and do the Wtxn call in `hashJoinStart`
- call the Close method on the account in joinNode's Close method instead of buckets'.

---

*[pkg/sql/row_container.go, line 197 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWmtYu3O67LFV_UqhMd:-KWmtYu3O67LFV_UqhMe:b-u3bb48) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1efe00e0f38bf295623bf0e4ba3f22a586a109a/pkg/sql/row_container.go#L197)):*
> ```Go
> 	copy(c.chunks[chunk][pos:pos+c.numCols], row)
> 	c.numRows++
> 	return c.chunks[chunk][pos : pos+c.numCols : pos+c.numCols], nil
> ```

Nice idea. I like this!

---

*[pkg/sql/row_container.go, line 31 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWmuPohVMePSeoQw4Dz:-KWmuPoitNqC3uGTKR0G:b-dmwc7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d74aa090b918f268b3b2705f7f8bb9fd86b31ad9/pkg/sql/row_container.go#L31)):*
> ```Go
> 	// targetChunkSize is the target number of Datums in a RowContainer chunk.
> 	targetChunkSize = 64
> 	sizeOfDatum     = int64(unsafe.Sizeof(parser.Datum(nil)))
> ```

Very nice, thanks for doing this!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWn1IhHDWjVzHqq1R_q:b-ezofag)*
<!-- Sent from Reviewable.io -->

knz commented:
Well done, btw. I dig this PR :) :yellow_heart: 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWn20SuUzAjwcS6yBvq:be3xk9b)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
Thank you for letting me work on this, I did give you a lot to review as well so thank you for being so prompt with that!

---

Review status: 18 of 19 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/join.go, line 76 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-KWn1K9CJZSqpUNP8k9R:-KWn4XoewX439BX73zDa:bzdwjv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d74aa090b918f268b3b2705f7f8bb9fd86b31ad9/pkg/sql/join.go#L76)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> - Add the WrappableMemoryAccount in joinNode.
> - pass WrappedMemoryAccount as argument to `buckets.AddRow`, and do the Wtxn call in `hashJoinStart`
> - call the Close method on the account in joinNode's Close method instead of buckets'.
> 
</details>
Great suggestion, looks much cleaner! Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWn4brHbWB2XS_zJ38b:b-66yhhc)*
<!-- Sent from Reviewable.io -->

knz commented:
:+1:

---

Reviewed 1 of 2 files at r10, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10618#-:-KWn7xSlIHWh_dg8lBPh:bb6mt9b)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Great stuff!! 👍
